[APP-243] feat: revive monorepo

dalechyn opened the pull request:
This is work in progress.
The proposed monorepo change would allow us ship both CJS, ESM and DTS files split in different proxy packages for better tree-shacking purposes.

As well as it "fixes" (and ignores) some of the issues within the `core` package, and yet needs more type-safety fixes on `react` package.

Implements APP-243 (https://linear.app/abstract-sdk/issue/APP-243/refactor-the-current-sdk)

dalechyn committed:
feat: revive monorepo

This is work in progress.
The proposed monorepo change would allow us ship both CJS, ESM and DTS
files split in different proxy packages for better tree-shacking
purposes.

As well as it "fixes" (and ignores) some of the issues within the `core`
package, and yet needs more type-safety fixes on `react` package.

dalechyn committed:
fix: build

dalechyn committed:
fix: types

dalechyn committed:
ci: drop test typecheck

dalechyn commented:
**Size** job fails but should be operating normally after the merge.

Post-action bug in there.

adairrr committed:
Remove internal-only functions

dalechyn committed:
feat: example app

dalechyn committed:
feat: add mutations

Commit:
chore: format

dalechyn committed:
feat: create round wip example

Commit:
chore: format

adairrr commented on examples/wagemos-app-nextjs/src/app/_components/connect-button.tsx:
```diff
@@ -0,0 +1,81 @@
+  const [isOpen, setIsOpen] = useState(false)
+  const { toast } = useToast()
+
+  const { isConnected, isConnecting } = useAccount({
```
Oof so both graz and abstract have `useAccount` that's not amazing...

dalechyn replied:
yep, lets rename ours

dalechyn replied:
can't find any `useAccount` hook from our side, resolving this thread.

adairrr commented on examples/wagemos-app-nextjs/src/generated/wagemos/index.ts:
```diff
@@ -0,0 +1,752 @@
+          Module: WagemosAppQueryClient,
+        },
+        { enabled: options?.enabled },
+      )
```
Perhaps we should also pull this into an autogenerated hook `useWagemosQueryClient`

dalechyn replied:
could you describe a bit more what do you mean?

adairrr commented on examples/wagemos-app-nextjs/src/generated/wagemos/index.ts:
```diff
@@ -0,0 +1,752 @@
+        isError: false,
+        isSuccess: true,
+      } as const
+    },
```
Nice! We should be able to pull this out into a function as well I'd assume...

dalechyn replied:
yep, definitely

dalechyn replied:
tried making a function, couldn't make it typesafe.
let's leave it as it is and improve incrementally.

adairrr commented on packages/core/src/__generated__/gql/index.ts:
```diff
@@ -0,0 +1,2 @@
+export * from './gql'
+export * from './fragment-masking'
```
We should figure out a way to ensure that these generated files don't end up in the published version

dalechyn replied:
should granularly export everything instead

dalechyn replied:
i won't refactor the core sdk logic itself unless it produces bad bundles, therefore won't touch it.

adairrr commented on packages/core/src/api/apiRequest.ts:
```diff
@@ -0,0 +1,31 @@
+ * @param gqlQuery
+ * @param variablesAndRequestHeaders
+ */
+export const abstractApiRequest = <TResult, V extends Variables = Variables>(
```
We'll need to update these from the monorepo once more 😅

dalechyn replied:
lets add a todo

dalechyn replied:
updated to match.

adairrr commented on packages/core/src/clients/AnsClient.ts:
```diff
@@ -194,8 +217,6 @@ interface IAnsClient extends IAnsQueryClient, AnsHostInterface {}
-    super(client, sender, contractAddress)
-  }
-}
+export class AnsClient
+  extends Mixin(AnsHostClient, AnsQueryClient)
+  implements IAnsClient {}
```
Does this still function without the constructor?

dalechyn replied:
yep, it will take a constructor from the successor class

adairrr commented on packages/core/codegen.yml:
```diff
@@ -0,0 +1,12 @@
+schema: "../../apps/subgraph/schema.graphql"
```
We'll need to update this. Perhaps we should use a ts config and load from env?

dalechyn replied:
replaced the schema url. yml files don't support env variables and since it's a library that most probably won't need another subgraph scheme let's have it hardcoded.

if any contributors would like to change the schema we should let them know that they will have to replace the schema url although it seems that it should be implicitly trivial.

adairrr commented on packages/react/src/contexts/account-id.ts:
```diff
@@ -0,0 +1,39 @@
+ * consumer is not obligated to set one at the initialization time, the context
+ * is exported as a separate entity to be used whenever needed.
+ */
+export function AbstractAccountIdContext({
```
Nice!

adairrr commented on packages/react/src/contexts/provider.ts:
```diff
@@ -0,0 +1,37 @@
+  },
+})
+
+export type AbstractProviderProps = AbstractConfigProps
```
We should probably allow the consumer to provide query client options as well - this is what graz does https://github.com/graz-sh/graz/blob/2dd73ddae3313e76be5f79215a30c0916c9c1fcd/packages/graz/src/provider/index.tsx#L2

dalechyn replied:
or a client should instead init the query client on his side and inject it into the provider, we could build a fancy `createConfig` as wagmi did.

adairrr commented on packages/react/src/utils/use-module-mutation-client.ts:
```diff
@@ -0,0 +1,163 @@
+import { SigningCosmWasmClient } from '@cosmjs/cosmwasm-stargate'
+import { useAccount, useCosmWasmSigningClient, useRecentChains } from 'graz'
```
This ties the consumers to use graz... while the most commonplace provider is cosmos-kit. Perhaps we should pull out the hooks that we would use into our own react hooks?

adairrr replied:
Unless we also provide the graz context underneath...

dalechyn replied:
fixed

dalechyn committed:
feat: decouple graz from sdk and use in generated module

dalechyn committed:
feat: mutation examples

Also commented out the dev build for the  package, since nextjs
example tries to transpile it and requries dependencies to be installed.

dalechyn committed:
wip: build issues

Unfortunately I am meeting building issues again.

dalechyn committed:
fix: core package build

Turned out to be a bad remapping to the package path.

dalechyn committed:
wip: build issues

Still having return type annotations issues.

dalechyn committed:
fix lint issues

dalechyn committed:
fix: build

It turned out that `tsup`s `dts` flag has been generating
incorrect typings, so I tried using `experimentalDts` and it
worked.

`react` package won't use this option though, since it doesn't
work there.
DTS building is a wild world.

dalechyn committed:
fix: linearize build with turbo

dalechyn committed:
fix: add dependencies for typecheck job

dalechyn commented:
@adairrr lint, build, types jobs pass.
PR is ready for the review.

akaia-shadowfox commented on .vscode/extensions.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "recommendations": ["orta.vscode-twoslash-queries", "rome.rome"]
```
Since Rome has changed ownership and rebranded as Biome, there's the new extension: https://marketplace.visualstudio.com/items?itemName=biomejs.biome

dalechyn replied:
oh sure, will change.

akaia-shadowfox commented on .vscode/settings.json:
```diff
@@ -0,0 +1,25 @@
+    "source.organizeImports.rome": true
+  },
+  "[json]": {
+    "editor.defaultFormatter": "rome.rome"
```
( See my previous comment )

dalechyn committed:
chore: update recommended linter for VSCode to `biome`

dalechyn committed:
chore: update settings for VSCode to use biome

dalechyn committed:
nit: updated `apiRequest` to match the frontend monorepo

https://github.com/AbstractSDK/frontend/blob/68b8d9b17993211dfa54a2674bf7da4991d78d83/packages/abstractjs/src/api/apiRequest.ts

dalechyn committed:
fix: replace schema url for codegen

adairrr commented on packages/react/src/utils/use-module-query-client.ts:
```diff
@@ -0,0 +1,147 @@
+  moduleId: string
+  Module: TModule
+}) {
+  // TODO: re-check if grabbing the first chain of the list is a good solution
+  const data = await graphqlRequest(overrideApiUrl, CHAIN_DEPLOYMENT_QUERY, {
+    chain,
+  })
```
It's not a good solution - we'll need to have the chainName passed in as a parameter or provided in some sort of context

dalechyn replied:
Noted at https://linear.app/abstract-sdk/issue/APP-248/fix-the-chain-name-assertion-in-pseudogenerated-modules

adairrr commented on examples/wagemos-app-nextjs/src/generated/wagemos.graz/index.ts:
```diff
@@ -0,0 +1,799 @@
+  options?: Parameters<typeof useModuleQueryClient>[1],
+) {
+  const { data: client } = useCosmWasmClient()
+  const chainIds = useActiveChainIds()
+  const chain = chainIds?.[0]
+  return useModuleQueryClient(
```
As this provides the query and execution logic for that specific chain, we can't safely assume that it's going to be the first one, so we may have to have chain as a parameter or context.

dalechyn replied:
Noted at https://linear.app/abstract-sdk/issue/APP-248/fix-the-chain-name-assertion-in-pseudogenerated-modules

adairrr commented on examples/wagemos-app-nextjs/src/app/@bets/_components/place-bet-dialog.tsx:
```diff
@@ -0,0 +1,164 @@
+  const { mutateAsync: placeBetAsync, isLoading } =
+    wagemos.mutations.usePlaceBet()
+
+  const onSubmit: SubmitHandler<z.infer<typeof placeBetSchema>> = useCallback(
+    async ({ amount, accountSeq }) => {
+      const accountIdToBetOn = round.teams.find(
+        (account) => account.seq === +accountSeq,
+      )
+      if (!accountIdToBetOn) throw new Error('Account not found')
+
+      await placeBetAsync?.({
+        msg: {
+          roundId: round.id,
+          bet: {
+            account_id: accountIdToBetOn,
+            asset: { name: round.total_bet.name, amount: amount.toString() },
+          },
+        },
+      })
+
+      setIsOpen(false)
+      toast({
+        title: 'Bet placed',
+        description: `You bet ${amount} ${round.total_bet.name} on account ${accountIdToBetOn.seq}`,
+      })
+    },
+    [placeBetAsync],
+  )
```
nit small design question :
 
 Is this generally the best way to deal with mutations in react-query? From what I've seen we want to avoid using the async mutation and add additinoal mutations in the components with the onSuccess and onError handlers

dalechyn replied:
While react-query allows both, I assume it is safe to use both variants unless the post-mutation logic is triggered more than once.

adairrr reviewed (approved):
Let's get this merged and we can fix the remaining chain comments afterwards!
